pilot/pkg/config/memory/monitor.go

```diff
@@ -77,27 +80,32 @@ func newBufferedMonitor(store model.ConfigStore, bufferSize int, sync bool) Moni
 }
 
 func (m *configStoreMonitor) ScheduleProcessEvent(configEvent ConfigEvent) {
+	if m.closed.Load() {
+		return
+	}
+
 	if m.sync {
 		m.processConfigEvent(configEvent)
 	} else {
 		m.eventCh <- configEvent
 	}
 }
 
+func (m *configStoreMonitor) run(stop <-chan struct{}) {
+	<-stop
+	m.closed.Store(true)
+	close(m.eventCh)
```

Comment:
Why close it? I think this can cause a panic, if another thread is executing L90 at the same time. https://stackoverflow.com/a/8593986/374797

Reply:
Thank you for your review, i think the channel will lead to memory leaks before. Can i use an empty event to indicates the exiting of the `run` function

Reply:
what about when Stop is received, we loop through all remaining `eventCh`?

Actually, neither the current approach nor that will work. `ScheduleProcessEvent` maybe have 100 goroutines on  L86 that send to eventCh after stop is closed which would be missed?

Reply:
Year,  loop through all remaining `eventCh` is necessary, what about the code snippet and add it to the end `Run` function.
```
for len(m.eventCh) > 0 {
    ce := <-m.eventCh
    m.processConfigEvent(*ce)
}
```

Reply:
I have another idea, can we add a lock to make sure the `closed` and `eventCh` operator is atomic. such as 
```
func (m *configStoreMonitor) ScheduleProcessEvent(configEvent *ConfigEvent) {
	Lock()
	if m.closed.Load() || configEvent == nil {
		Unlock()
		return
	}

	if m.sync {
		Unlock()
		m.processConfigEvent(*configEvent)
	} else {
		Unlock()
		m.eventCh <- configEvent
	}
}

func (m *configStoreMonitor) run(stop <-chan struct{}) {
	<-stop
	Lock()
	m.closed.Store(true)
	// indicating the monitor has exited.
	m.eventCh <- nil
	Unlock()
}
```

Reply:
I guess my overall question - why do we need this PR at all. Does it matter if we drop events? After all, we closed the store anyways

Reply:
The customer(such as https://github.com/istio/istio/blob/cced4dfa0f86cf18efbfa0daefe08b09e36cdb44/pilot/pkg/serviceregistry/serviceentry/controller.go#L143) of the events tracks the change of store. If the events lost and the lifecycle of customer and store are not same, that may cause unexpected behavior.

Reply:
Is there any real world case where this impacts Istio users?

Reply:
There is no case in the istio now as the store's lifecycle is same as pilotd. I meet it in our project which has some custom feature.

Reply:
The featur is about add or remove the [ads](https://github.com/istio/istio/blob/539a27879d63e5186e13e5781c20e0ed4c4c7d3f/pkg/adsc/adsc.go#L143) client dynamically, the ads client use the storeMonitor.